### PR TITLE
Restructure datatype enforcement code.

### PR DESCRIPTION
**Pull request by arh89**
To-do:
* Add docstrings.
* Possibly improve unit test coverage?
* Review which functions are public/rename functions etc (might want a better interface).

### CODE REVIEW

File: improver/metadata/check_datatypes.py [hunk 1]
```diff
@@ -30,201 +30,85 @@
...
+    dtype_ok = obj.dtype == req_dtype
+
+    if isinstance(obj, iris.coords.Coord) and obj.has_bounds():
+        # check bounds - want the same dtype as the points
+        bounds_dtype_ok = obj.bounds.dtype == req_dtype
+        dtype_ok = dtype_ok and bounds_dtype_ok
+    return dtype_ok
+
+
+def get_required_units(obj):
```
Comment (benfitzpatrick) [inline improver/metadata/check_datatypes.py]: Codacy Issue found: [Missing function or method docstring](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4882613)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 6
           

Complexity decreasing per file
==============================
+ improver/metadata/check_datatypes.py  -5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4882613&bid=16178027)

File: improver/standardise.py [hunk 2]
```diff
@@ -267,9 +234,26 @@ def _remove_scalar_coords(cube, coords_to_remove):
...
+            req_dtype = get_required_dtype(coord)
+            # ensure points and bounds have the same dtype
+            coord.points = as_correct_dtype(coord.points, req_dtype)
+            if coord.has_bounds():
+                coord.bounds = as_correct_dtype(coord.bounds, req_dtype)
+
     def process(self, cube, target_grid=None, new_name=None, new_units=None,
                 regridded_title=None, coords_to_remove=None,
-                attributes_dict=None, fix_float64=False):
+                attributes_dict=None, standardise_dtypes=False):
```
Comment (benfitzpatrick) [inline improver/standardise.py]: Codacy Issue found: [Unused argument 'standardise_dtypes'](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4882613)

File: improver/metadata/check_datatypes.py [hunk 3]
```diff
@@ -30,201 +30,85 @@
...
+    if _is_time_coord(obj):
+        return TIME_COORDS[obj.name()].units
+    # No special requirement on non-time coordinate objects
+    return None
+
+
+def check_units(obj):
```
Comment (benfitzpatrick) [inline improver/metadata/check_datatypes.py]: Codacy Issue found: [Missing function or method docstring](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4882613)

File: improver/metadata/check_datatypes.py [hunk 4]
```diff
@@ -30,201 +30,85 @@
...
-        if not _check_units_and_dtype(coord, required_unit, required_dtype):
-            msg = ('Coordinate {} does not match required '
-                   'standard (units {}, datatype {})\n')
-            error_string += msg.format(
-                coord.name(), required_unit, required_dtype)
-
-    # if non-compliance was encountered, raise all messages here
-    if error_string:
-        raise ValueError(error_string)
+def check_datatypes(cube):
```
Comment (benfitzpatrick) [inline improver/metadata/check_datatypes.py]: Codacy Issue found: [Missing function or method docstring](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4882613)

File: improver/metadata/check_datatypes.py [hunk 5]
```diff
@@ -30,201 +30,85 @@
...
-            continue
+from improver.metadata.constants import FLOAT_DTYPE
+from improver.metadata.constants.time_types import TIME_COORDS
+
+
+def _is_time_coord(obj):
+    return isinstance(obj, iris.coords.Coord) and obj.name() in TIME_COORDS
+
+
+def get_required_dtype(obj):
```
Comment (benfitzpatrick) [inline improver/metadata/check_datatypes.py]: Codacy Issue found: [Missing function or method docstring](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4882613)

File: improver/metadata/check_datatypes.py [hunk 6]
```diff
@@ -30,201 +30,85 @@
...
-        bool:
-            True if object conforms; False if not
-    """
-    if Unit(obj.units) != Unit(units):
-        return False
-
-    if obj.dtype.type != dtype:
-        return False
 
+def check_dtype(obj):
```
Comment (benfitzpatrick) [inline improver/metadata/check_datatypes.py]: Codacy Issue found: [Missing function or method docstring](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4882613)

File: improver/metadata/check_datatypes.py (hunk 3 above)
Comment (arh89) [inline improver/metadata/check_datatypes.py]: Probably leave `check_datatypes` as the name of the below function, but rename this to `valid_units` (and `valid_dtype` above)?

**Comment by MoseleyS** [comment]
Replaced by #1143

**Comment by arh89** [comment]
Thanks, @MoseleyS 👍

**Pull request closed without merging**